### PR TITLE
Install pg_attribute_encoding for AOCS during CTAS

### PR DESCRIPTION
**Pull request by x4m**
When the table is created using "CREATE TABLE ... AS ... WITH NO DATA" pg_attribute_encoding may be omitted in some cases. This can result in a crash when reading the table with AOCS storage.
This commit fixes installment of pg_attribute_encoding for CTAS WITH NO DATA and adds a test for the case.
Fix for #12936

### CODE REVIEW

File: src/backend/commands/createas.c [hunk 1]
```diff
@@ -307,12 +307,10 @@ create_ctas_nodata(List *tlist, IntoClause *into, QueryDesc *queryDesc)
 	intoRelationOid = create_ctas_internal(attrList, into, queryDesc, true);
 
 	/* Add column encoding entries based on the WITH clause */
-	if (into->options)
-	{
-		Relation rel = heap_open(intoRelationOid, AccessExclusiveLock);
+	Relation rel = heap_open(intoRelationOid, AccessExclusiveLock);
+	if (into->options || RelationIsAoCols(rel))
```
Comment (ashwinstar) [inline src/backend/commands/createas.c:311]: I don't think we even need this `if` condition at all. Inside `AddDefaultRelationAttributeOptions()` check exist for `RelationIsAoCols()` or not. Plus, based on modified code `into->options` existence or not makes no difference to calling `AddDefaultRelationAttributeOptions()`

File: src/backend/tcop/utility.c [hunk 2]
```diff
@@ -1168,7 +1168,7 @@ ProcessUtilitySlow(Node *parsetree,
 							CommandCounterIncrement();
 
 							/* Add column encoding entries based on the WITH clauses */
-							if (cstmt->isCtas && cstmt->options)
+							if (cstmt->isCtas && (cstmt->options || relStorage == RELSTORAGE_AOCOLS))
```
Comment (ashwinstar) [inline src/backend/tcop/utility.c:1171]: Please can you add test which exposes this code path and validates the fix for it? Other change is getting exercised via currently added test.

File: src/test/regress/sql/gpctas.sql [hunk 3]
```diff
@@ -160,3 +160,11 @@ from
 
 -- then refresh should error out
 refresh materialized view sro_mv_issue_11999;
+
+
+-- CTAS into AOCS table with not crash
+-- See github issue: https://github.com/greenplum-db/gpdb/issues/12936
+
+set gp_default_storage_options to 'appendonly=true,blocksize=32768,compresstype=zstd,compresslevel=1,checksum=true,orientation=column';
+CREATE TABLE test_issue_12936 AS (SELECT 'test' as test) WITH NO DATA;
```
Comment (ashwinstar) [inline src/test/regress/sql/gpctas.sql:169]: we should validate the table properties to make sure its created properly with desired   `gp_default_storage_options` or not. Current test is only validating if can read the data or not. What if the underlying table is created as heap :-)

File: src/backend/tcop/utility.c (hunk 2 above)
Comment (x4m) [reply src/backend/tcop/utility.c:1171]: Without this change segments fail the test.

**Comment by x4m** [comment]
@ashwinstar can you please look into what's up with Concourse CI? I can't open it.

**Comment by x4m** [comment]
Thanks, @kainwen! Will fix.. 
Seems like something wrong with my local test runs..

**Comment by x4m** [comment]
@kainwen can I restart Concourse CI tests

**Comment by x4m** [comment]
Any news here?

**Comment by kainwen** [comment]
Sorry for late.

Some code are to fix https://github.com/greenplum-db/gpdb/issues/9790.

And master branch changes more: https://github.com/greenplum-db/gpdb-postgres-merge/commit/2cdd21b826df46

Seems this PR lack at least one change in `open_ds_write`.

I'll spend more time to understand all the context and history and then to review this. Will do it soon.

Thanks for the PR.

**Comment by SmartKeyerror** [comment]
Hi @x4m , thanks for your contribution! This pr is looked fine for me, but we can do better:

Due to the commit: [Turn a couple of Asserts to elog()s.](https://github.com/greenplum-db/gpdb-postgres-merge/commit/2cdd21b826df46), I think it's better to add runtime checks in `open_ds_write()` as this PR did. Although we cannot easily construct badcases for write condition currently, but it'll improve the robustness of the code: if there are some bugs or catalog corruption, QE will not PANIC and EXIT, it just output friendly error message like "No relation attribute options for ..., column ...".

Another suggestion is to make the commit message more detailed. Such like:

- What this PR did.
- What problem this PR resolved.

A detailed commit message makes things easier :)

Thanks a lot.

**Comment by x4m** [comment]
@SmartKeyerror I've added Heikki's commit to this PR.
Would you prefer me to composing commit message here or squashing all commits to one with proper commit message?

**Review by SmartKeyerror — APPROVED** [review]
my local full pipelines are all green. After approved by ship-it meeting, we can merge it.

**Comment by x4m** [comment]
BTW when is the next Ship-it meeting scheduled?

**Comment by kainwen** [comment]
> BTW when is the next Ship-it meeting scheduled?

Tomorrow

**Comment by kainwen** [comment]
> > BTW when is the next Ship-it meeting scheduled?
> 
> Tomorrow

Due to public holiday in the US, let's wait for one more day.

**Pull request merged**